compiler/rustc_target/src/abi/mod.rs

```diff
@@ -2,1315 +2,32 @@ pub use Integer::*;
-            if use_small {
-                for i in 0..self.count() {
-                    inverse_small[memory_index[i] as usize] = i as u8;
-                }
-            } else {
-                inverse_big = vec![0; self.count()];
-                for i in 0..self.count() {
-                    inverse_big[memory_index[i] as usize] = i as u32;
-                }
-            }
-        }
-
-        (0..self.count()).map(move |i| match *self {
-            FieldsShape::Primitive | FieldsShape::Union(_) | FieldsShape::Array { .. } => i,
-            FieldsShape::Arbitrary { .. } => {
-                if use_small {
-                    inverse_small[i] as usize
-                } else {
-                    inverse_big[i] as usize
-                }
-            }
-        })
-    }
-}
-
-/// An identifier that specifies the address space that some operation
-/// should operate on. Special address spaces have an effect on code generation,
-/// depending on the target and the address spaces it implements.
-#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
-pub struct AddressSpace(pub u32);
-
-impl AddressSpace {
-    /// The default address space, corresponding to data space.
-    pub const DATA: Self = AddressSpace(0);
-}
-
-/// Describes how values of the type are passed by target ABIs,
-/// in terms of categories of C types there are ABI rules for.
-#[derive(Clone, Copy, PartialEq, Eq, Hash, Debug, HashStable_Generic)]
-pub enum Abi {
-    Uninhabited,
-    Scalar(Scalar),
-    ScalarPair(Scalar, Scalar),
-    Vector {
-        element: Scalar,
-        count: u64,
-    },
-    Aggregate {
-        /// If true, the size is exact, otherwise it's only a lower bound.
-        sized: bool,
-    },
-}
-
-impl Abi {
-    /// Returns `true` if the layout corresponds to an unsized type.
-    #[inline]
-    pub fn is_unsized(&self) -> bool {
-        match *self {
-            Abi::Uninhabited | Abi::Scalar(_) | Abi::ScalarPair(..) | Abi::Vector { .. } => false,
-            Abi::Aggregate { sized } => !sized,
-        }
-    }
-
-    #[inline]
-    pub fn is_sized(&self) -> bool {
-        !self.is_unsized()
-    }
-
-    /// Returns `true` if this is a single signed integer scalar
-    #[inline]
-    pub fn is_signed(&self) -> bool {
-        match self {
-            Abi::Scalar(scal) => match scal.primitive() {
-                Primitive::Int(_, signed) => signed,
-                _ => false,
-            },
-            _ => panic!("`is_signed` on non-scalar ABI {:?}", self),
-        }
-    }
-
-    /// Returns `true` if this is an uninhabited type
-    #[inline]
-    pub fn is_uninhabited(&self) -> bool {
-        matches!(*self, Abi::Uninhabited)
-    }
-
-    /// Returns `true` is this is a scalar type
-    #[inline]
-    pub fn is_scalar(&self) -> bool {
-        matches!(*self, Abi::Scalar(_))
-    }
-}
-
 rustc_index::newtype_index! {
     pub struct VariantIdx {
         derive [HashStable_Generic]
     }
 }
 
```

Comment:
Why keep this here? Does RA want to substitute its own?

(Long-term it might make sense to have `Layout` take, as a generic parameter, some type that implements a trait with one associated parameter per type we want to customize etc. - instead of just adding more and more parameters - if the need arises, anyway)

Speaking of which, such a generic parameter would've allowed you to keep interning for variants, for rustc (which IIRC @nnethercote added intentionally), but use e.g. `Box<Layout<...>>` in other cases.

Reply:
> Why keep this here? Does RA want to substitute its own?

It needed to make changes rustc_macros, and r-a had an index system itself, so I choosed to do this.

> Long-term it might make sense to have Layout take, as a generic parameter, some type that implements a trait with one associated parameter per type

Similar to chalk's interner. It make sense, but currently there is only one such argument. If we want to bring interning of nested layouts back (which doesn't showed itself in perf, maybe it only affects memory usage?) I can do that.